solarforecastarbiter/datamodel.py

```diff
@@ -934,6 +938,231 @@ def __post_init__(self):
+       The datetimes to associate with each cost value
+    cost : tuple of float
+       The cost per unit error of the forecasted variable for each datetime.
+       Must have the same length as `datetimes`.
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    fill : str
+       Fill method to apply for datetimes between those specified in
+       `datetimes`. Options are 'forward' or 'backward'.
+    timezone : str, default None
+        IANA timezone string to use when constructing datetimes. If None,
+        the timezone of the observations is used, which is the report
+        timezone when calculated in a report.    """
+    datetimes: Tuple[pd.Timestamp, ...]
+    cost: Tuple[float, ...]
+    aggregation: str
+    net: bool
+    fill: str
+    timezone: str = None
+    __post_init__ = __validate_cost__('datetimes')
+
+
+@dataclass(frozen=True)
+class ConstantCost(BaseModel):
+    """A constant cost per unit error of the forecasted variable
+
+    Parameters
+    ----------
+    cost : float
+    aggregation : str
+        Aggregation method to use after calculating cost for the error series.
+        Currently only 'sum' or 'mean' are available.
+    net : bool
+        If True, compute the 'net' aggregate error instead of first calcuating
+        the absolute error before performing the aggregation.
+    """
+    cost: float
+    aggregation: str
+    net: bool
+    __post_init__ = __validate_cost__(None)
+
+
+@dataclass(frozen=True)
+class CostBand(BaseModel):
+    """Cost specification for one error band
+
+    Parameters
+    ----------
+    error_range : tuple(float, float)
+        Bounds of the error to apply the specified cost function to.
+        Inf and -Inf are valid range points, and the error may be positive or
+        negative. Inclusion/exclusion of endpoints is determined by ordering
+        in :py:class:`solarforecastarbiter.datamodel.ErrorBandCost`.
+    cost_function : str
+        One of 'timeofday', 'datetime', or 'constant'. Specifies which
+        cost model should be used to calculate the cost in this band.
+    cost_function_parameters : :py:class:`solarforecastarbiter.datamodel.ConstantCost` or :py:class:`solarforecastarbiter.TimeOfDayCost` or :py:class:`solarforecastarbiter.DatetimeCost`
+        Parameters for the selected cost function.
+    """  # NOQA
+    error_range: Tuple[float, float]
+    cost_function: str
+    cost_function_parameters: Union[TimeOfDayCost, DatetimeCost, ConstantCost]
+
+    def _special_field_processing(self, model_field, val):
+        # support passing "inf", "-inf" as strings via json/dict
+        if model_field.name == 'error_range':
+            return float(val)
```

Comment:
so far as I can tell, this line is only hit because of how the test report is configured. maybe should add a datamodel-specific test through the json in conftest.

Reply:
Sure